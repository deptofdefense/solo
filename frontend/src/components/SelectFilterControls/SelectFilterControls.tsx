import React, { useState } from "react";
import classNames from "classnames";
import { Button } from "solo-uswds";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSearch } from "@fortawesome/free-solid-svg-icons";
import classes from "./SelectFilterControls.module.scss";

interface SelectFilterControls {
  onSubmit: (filter: { option: string; value: string }) => void;
  options: Array<{ name: string; value: string }>;
  defaultOption?: string;
}

const SelectFilterControls: React.FC<SelectFilterControls> = ({
  options,
  defaultOption,
  onSubmit
}) => {
  const [currentOption, setCurrentOption] = useState(
    defaultOption ?? options[0]?.value
  );
  const [currentValue, setCurrentValue] = useState("");

  const onOptionChange: React.ChangeEventHandler<HTMLSelectElement> = event => {
    setCurrentOption(event.currentTarget.value);
  };

  const onValueChange: React.ChangeEventHandler<HTMLInputElement> = event => {
    setCurrentValue(event.currentTarget.value);
  };

  const onSubmitted: React.FormEventHandler<HTMLFormElement> = event => {
    event.preventDefault();
    onSubmit({
      option: currentOption,
      value: currentValue
    });
  };

  return (
    <form
      className="grid-row flex-row flex-justify-center flex-align-center"
      onSubmit={onSubmitted}
    >
      <div className="grid-col-full tablet:grid-col-2 margin-x-2">
        <select
          onChange={onOptionChange}
          value={currentOption}
          className="usa-select"
          placeholder="Field"
        >
          {options.map(({ name, value }) => (
            <option key={value} value={value}>
              {name}
            </option>
          ))}
        </select>
      </div>
      <div className="grid-col-10 tablet:grid-col-3">
        <input
          value={currentValue}
          onChange={onValueChange}
          className={classNames("usa-input", classes.select)}
          name="input-type-text"
          type="text"
          placeholder="Search"
        />
      </div>
      <div className="grid-col-auto">
        <Button square className="margin-top-1" type="submit">
          <FontAwesomeIcon icon={faSearch} />
        </Button>
      </div>
    </form>
  );
};

export default SelectFilterControls;
