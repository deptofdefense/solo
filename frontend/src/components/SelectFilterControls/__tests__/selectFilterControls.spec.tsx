import React from "react";
import { render, fireEvent, wait } from "test-utils";
import SelectFilterControls from "../SelectFilterControls";

describe("SelectFilterControls component", () => {
  const testOpts = [
    { name: "test opt 1", value: "testopt1" },
    { name: "test opt 2", value: "testopt2" },
    { name: "test opt 3", value: "testopt3" }
  ];
  const submitCallback = jest.fn();

  afterEach(() => {
    submitCallback.mockReset();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <SelectFilterControls options={testOpts} onSubmit={submitCallback} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("includes all passed options", () => {
    const { queryByText } = render(
      <SelectFilterControls options={testOpts} onSubmit={submitCallback} />
    );
    expect(queryByText("test opt 1")).toBeInTheDocument();
    expect(queryByText("test opt 2")).toBeInTheDocument();
    expect(queryByText("test opt 3")).toBeInTheDocument();
  });

  it("sets current field based on default option when rendered", async () => {
    const { getByPlaceholderText } = render(
      <SelectFilterControls
        options={testOpts}
        onSubmit={submitCallback}
        defaultOption="testopt3"
      />
    );
    const select = getByPlaceholderText("Field");
    expect(select).toHaveValue("testopt3");
  });

  it("calls submit callback with field and value on form submit", async () => {
    const { getByPlaceholderText, container } = render(
      <SelectFilterControls
        options={testOpts}
        onSubmit={submitCallback}
        defaultOption="testopt2"
      />
    );
    const btn = container.querySelector("button");
    const search = getByPlaceholderText("Search");
    fireEvent.change(search, {
      target: {
        value: "my search value"
      }
    });
    await wait(() => {
      expect(search).toHaveValue("my search value");
    });
    fireEvent.click(btn as Element);
    await wait(() => {
      expect(submitCallback).toHaveBeenCalled();
      expect(submitCallback.mock.calls[0][0]).toMatchObject({
        option: "testopt2",
        value: "my search value"
      });
    });
  });
});
