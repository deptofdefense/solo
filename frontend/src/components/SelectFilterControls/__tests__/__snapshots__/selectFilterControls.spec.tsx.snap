// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectFilterControls component matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="grid-row flex-row flex-justify-center flex-align-center"
  >
    <div
      class="grid-col-full tablet:grid-col-2 margin-x-2"
    >
      <select
        class="usa-select"
        placeholder="Field"
      >
        <option
          value="testopt1"
        >
          test opt 1
        </option>
        <option
          value="testopt2"
        >
          test opt 2
        </option>
        <option
          value="testopt3"
        >
          test opt 3
        </option>
      </select>
    </div>
    <div
      class="grid-col-10 tablet:grid-col-3"
    >
      <input
        class="usa-input noOutline"
        placeholder="Search"
        type="text"
        value=""
      />
    </div>
    <div
      class="grid-col-auto"
    >
      <button
        class="usa-button margin-top-1 square"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
